HeterogeneousCore/CUDAUtilities/interface/launch.h

```diff
@@ -0,0 +1,139 @@
+#ifndef HeterogeneousCore_CUDAUtilities_launch_h
+#define HeterogeneousCore_CUDAUtilities_launch_h
+
+#include <tuple>
+
+#include <cuda_runtime.h>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+/*
+ * `cudautils::launch` and `cudautils::launch_cooperative` are wrappers around
+ * the CUDA Runtime API calls to setup and call a CUDA kernel from the host.
+ *
+ * `kernel` should be a pointer to a __global__ void(...) function.
+ * `config` describe the launch configuration: the grid size and block size, the
+ *          dynamic shared memory size (default to 0) and the CUDA stream to use
+ *          (default to 0, the default stream).
+ * `args` are the arguments passed (by value) to the kernel.
+ *
+ *  Currently this is requires an extra copy to perform the necessary implicit
+ *  conversions and ensure that the arguments match the kernel function signature;
+ *  the extra copy could eventually be avoided for arguments that are already of
+ *  the exact type.
+ *
+ *  Unlike the `kernel<<<...>>>(...)` syntax and the `cuda::launch(...)` 
+ *  implementation from the CUDA API Wrappers, `cudautils::launch(...)` and 
+ *  `cudautils::launch_cooperative` can be called from standard C++ host code.
+ */
+
+namespace cudautils {
+
+  struct LaunchParameters {
+    dim3 gridDim;
+    dim3 blockDim;
+    size_t sharedMem;
+    cudaStream_t stream;
+
+    LaunchParameters(dim3 gridDim, dim3 blockDim, size_t sharedMem = 0, cudaStream_t stream = nullptr)
+        : gridDim(gridDim), blockDim(blockDim), sharedMem(sharedMem), stream(stream) {}
+
+    LaunchParameters(int gridDim, int blockDim, size_t sharedMem = 0, cudaStream_t stream = nullptr)
+        : gridDim(gridDim), blockDim(blockDim), sharedMem(sharedMem), stream(stream) {}
+  };
+
+  namespace detail {
+
+    template <typename T>
+    struct kernel_traits;
+
+    template <typename... Args>
+    struct kernel_traits<void(Args...)> {
+      static constexpr size_t arguments_size = sizeof...(Args);
+
+      using argument_type_tuple = std::tuple<Args...>;
+
+      template <size_t i>
+      using argument_type = typename std::tuple_element<i, argument_type_tuple>::type;
+    };
+
+    // fill an array with the pointers to the elements of a tuple
+    template <int I>
+    struct pointer_setter {
+      template <typename Tuple>
+      void operator()(void const* ptrs[], Tuple const& t) {
+        pointer_setter<I - 1>()(ptrs, t);
+        ptrs[I - 1] = &std::get<I - 1>(t);
+      }
+    };
+
+    template <>
+    struct pointer_setter<0> {
+      template <typename Tuple>
+      void operator()(void const* ptrs[], Tuple const& t) {}
+    };
+
+#if 0
```

Comment:
this should either be changed to `#if __cplusplus >= 201703L` and implemented, or removed.